lib/src/dense_set.rs

```diff
@@ -0,0 +1,1282 @@
+        let reg_index = get_reg_index(item);
+        let data_index = get_data_index(reg_index);
+        let offset = get_offset(reg_index);
+
+        if data_index >= self.data.len() {
+            for _ in self.data.len()..data_index + 1 {
+                self.data.push(0);
+            }
+        }
+        self.data[data_index] |= 1 << offset;
+    }
+    fn vec_insert(&mut self, item: Reg) {
+        self.reg_vec.push(item);
+    }
+
+    fn contains(&self, item: Reg) -> bool {
+        let reg_index = get_reg_index(item);
+        let data_index = get_data_index(reg_index);
+        let offset = get_offset(reg_index);
+
+        if data_index >= self.data.len() {
+            false
+        } else {
+            (1 & (self.data[data_index] >> offset)) != 0
+        }
+    }
+
+    fn from_vec(vec: Vec<Reg>) -> Self {
+        let mut res: BitSet = RegSet::empty();
+
+        for &reg in vec.iter() {
+            res.data_insert(reg);
+        }
+        res.reg_vec = vec;
+
+        res
+    }
+
+    fn to_vec(&self) -> Vec<Reg> {
+        return self.reg_vec.clone();
+    }
+
+    fn delete(&mut self, item: Reg) {
+        self.vec_delete(item);
+        self.data_delete(item);
+    }
+
+    fn vec_delete(&mut self, item: Reg) {
+        let mut reg_index = 0;
+        for &mut reg in self.reg_vec.iter_mut() {
+            if reg == item {
+                self.reg_vec.swap_remove(reg_index);
+                break;
+            }
+            reg_index += 1;
+        }
+    }
+
+    fn data_delete(&mut self, item: Reg) {
+        let reg_index = get_reg_index(item);
+        let data_index = get_data_index(reg_index);
+        let offset = get_offset(reg_index);
+
+        self.data[data_index] &= !(1 << offset);
+    }
+
+    fn is_empty(&self) -> bool {
+        for i in 0..self.data.len() {
+            if self.data[i] != 0 {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn intersect(&mut self, other: &Self) {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+        let mut res = Vec::<Reg>::new();
+
+        for &item in self.reg_vec.iter() {
+            if other.contains(item) {
+                res.push(item);
+            }
+        }
+        self.reg_vec = res;
+
+        for i in 0..smallest_set_size {
+            self.data[i] &= other.data[i];
+        }
+
+        for i in smallest_set_size..self.data.len() {
+            self.data[i] = 0;
+        }
+    }
+
+    fn union(&mut self, other: &Self) {
+        let smallest_set_size = cmp::min(self.data.len(), other.data.len());
+        let greatest_set_size = cmp::max(self.data.len(), other.data.len());
+
+        for &item in other.reg_vec.iter() {
+            if !self.contains(item) {
+                self.reg_vec.push(item);
+            }
+        }
+
+        for i in 0..smallest_set_size {
+            self.data[i] |= other.data[i];
+        }
+
+        if other.data.len() > self.data.len() {
```

Comment:
to make this more explicit, how about `if other.data.len() == greatest_set_size`?

Reply:
Indeed it looks better. So it's done 055a40e35ea38e175be07820ad38f2c18de32ea7. Thanks !